File: mojo/public/bindings/sample/mojom/sample_service.cc
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojom/sample_service.h"

#include "mojo/public/bindings/lib/message_builder.h"
#include "mojom/sample_service_internal.h"

namespace sample {
namespace {
const uint32_t kService_Frobinate_Name = 1;

#pragma pack(push, 1)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunused-private-field"
#endif
class Service_Frobinate_Params {
 public:
  static Service_Frobinate_Params* New(mojo::Buffer* buf) {
    return new (buf->Allocate(sizeof(Service_Frobinate_Params)))
        Service_Frobinate_Params();
  }

  void set_foo(Foo* foo) { foo_.ptr = foo; }
  void set_baz(bool baz) { baz_ = baz; }
  void set_port(mojo::Handle port) { port_ = port; }

  const Foo* foo() const { return foo_.ptr; }
  bool baz() const { return baz_; }
  mojo::Handle port() const {
    // NOTE: port is an optional field!
    return _header_.num_fields >= 3 ? port_ : mojo::kInvalidHandle;
  }

 private:
  friend class mojo::internal::ObjectTraits<Service_Frobinate_Params>;

  Service_Frobinate_Params() {
    _header_.num_bytes = sizeof(*this);
    _header_.num_fields = 3;
  }

  mojo::internal::StructHeader _header_;
  mojo::internal::StructPointer<Foo> foo_;
  uint8_t baz_ : 1;
  uint8_t _pad0_[3];
  mojo::Handle port_;
};
MOJO_COMPILE_ASSERT(sizeof(Service_Frobinate_Params) == 24,
                    bad_sizeof_Service_Frobinate_Params);

const uint32_t kServiceClient_DidFrobinate_Name = 0;

class ServiceClient_DidFrobinate_Params {
 public:
  static ServiceClient_DidFrobinate_Params* New(mojo::Buffer* buf) {
    return new (buf->Allocate(sizeof(ServiceClient_DidFrobinate_Params)))
        ServiceClient_DidFrobinate_Params();
  }

  void set_result(int32_t result) { result_ = result; }

  int32_t result() const { return result_; }

 private:
  friend class mojo::internal::ObjectTraits<ServiceClient_DidFrobinate_Params>;

  ServiceClient_DidFrobinate_Params() {
    _header_.num_bytes = sizeof(*this);
    _header_.num_fields = 3;
  }

  mojo::internal::StructHeader _header_;
  int32_t result_;
  uint8_t _pad0_[4];
};
MOJO_COMPILE_ASSERT(sizeof(ServiceClient_DidFrobinate_Params) == 16,
                    bad_sizeof_ServiceClient_DidFrobinate_Params);

#if defined(__clang__)
#pragma clang diagnostic pop
#endif
#pragma pack(pop)

}  // namespace

// static
Bar* Bar::New(mojo::Buffer* buf) {
  return new (buf->Allocate(sizeof(Bar))) Bar();
}

Bar::Bar() {
  _header_.num_bytes = sizeof(*this);
  _header_.num_fields = 3;
}

// static
Foo* Foo::New(mojo::Buffer* buf) {
  return new (buf->Allocate(sizeof(Foo))) Foo();
}

Foo::Foo() {
  _header_.num_bytes = sizeof(*this);
  _header_.num_fields = 10;
}

ServiceProxy::ServiceProxy(mojo::MessageReceiver* receiver)
    : receiver_(receiver) {
}

void ServiceProxy::Frobinate(const Foo* foo, bool baz, mojo::Handle port) {
  size_t payload_size =
      mojo::internal::Align(sizeof(Service_Frobinate_Params));
  payload_size += mojo::internal::ComputeSizeOf(foo);

  mojo::MessageBuilder builder(kService_Frobinate_Name, payload_size);

  // We now go about allocating the anonymous Frobinate_Params struct.  It
  // holds the parameters to the Frobinate message.
  //
  // Notice how foo is cloned. This causes a copy of foo to be generated
  // within the same buffer as the Frobinate_Params struct. That's what we
  // need in order to generate a contiguous blob of message data.

  Service_Frobinate_Params* params =
      Service_Frobinate_Params::New(builder.buffer());
  params->set_foo(mojo::internal::Clone(foo, builder.buffer()));
  params->set_baz(baz);
  params->set_port(port);

  // NOTE: If foo happened to be a graph with cycles, then Clone would not
  // have returned.

  // Next step is to encode pointers and handles so that messages become
  // hermetic. Pointers become offsets and handles becomes indices into the
  // handles array.
  mojo::Message message;
  mojo::internal::EncodePointersAndHandles(params, &message.handles);

  // Finally, we get the generated message data, and forward it to the
  // receiver.
  message.data = builder.Finish();

  receiver_->Accept(&message);
}

bool ServiceStub::Accept(mojo::Message* message) {
  switch (message->data->header.name) {
    case kService_Frobinate_Name: {
      Service_Frobinate_Params* params =
          reinterpret_cast<Service_Frobinate_Params*>(
              message->data->payload);

      if (!mojo::internal::DecodePointersAndHandles(params, *message))
        return false;

      Frobinate(params->foo(), params->baz(), params->port());
      break;
    }
  }
  return true;
}

ServiceClientProxy::ServiceClientProxy(mojo::MessageReceiver* receiver)
    : receiver_(receiver) {
}

void ServiceClientProxy::DidFrobinate(int32_t result) {
  size_t payload_size =
      mojo::internal::Align(sizeof(ServiceClient_DidFrobinate_Params));


  mojo::MessageBuilder builder(kServiceClient_DidFrobinate_Name, payload_size);

  ServiceClient_DidFrobinate_Params* params =
      ServiceClient_DidFrobinate_Params::New(builder.buffer());

  params->set_result(result);

  mojo::Message message;
  mojo::internal::EncodePointersAndHandles(params, &message.handles);

  message.data = builder.Finish();

  receiver_->Accept(&message);
}


bool ServiceClientStub::Accept(mojo::Message* message) {
  switch (message->data->header.name) {
    case kServiceClient_DidFrobinate_Name: {
      ServiceClient_DidFrobinate_Params* params =
          reinterpret_cast<ServiceClient_DidFrobinate_Params*>(
              message->data->payload);

      if (!mojo::internal::DecodePointersAndHandles(params, *message))
        return false;
      DidFrobinate(params->result());
      break;
    }

  }
  return true;
}

}  // namespace sample

namespace mojo {
namespace internal {

// static
size_t ObjectTraits<sample::Bar>::ComputeSizeOf(
    const sample::Bar* bar) {
  return sizeof(*bar);
}

// static
sample::Bar* ObjectTraits<sample::Bar>::Clone(
    const sample::Bar* bar, Buffer* buf) {
  sample::Bar* clone = sample::Bar::New(buf);
  memcpy(clone, bar, sizeof(*bar));
  return clone;
}

// static
void ObjectTraits<sample::Bar>::EncodePointersAndHandles(
    sample::Bar* bar, std::vector<Handle>* handles) {
}

// static
bool ObjectTraits<sample::Bar>::DecodePointersAndHandles(
    sample::Bar* bar, const Message& message) {
  return true;
}

// static
size_t ObjectTraits<sample::Foo>::ComputeSizeOf(
    const sample::Foo* foo) {
  return sizeof(*foo) +
      mojo::internal::ComputeSizeOf(foo->bar()) +
      mojo::internal::ComputeSizeOf(foo->data()) +
      mojo::internal::ComputeSizeOf(foo->extra_bars()) +
      mojo::internal::ComputeSizeOf(foo->name()) +
      mojo::internal::ComputeSizeOf(foo->files());
}

// static
sample::Foo* ObjectTraits<sample::Foo>::Clone(
    const sample::Foo* foo, Buffer* buf) {
  sample::Foo* clone = sample::Foo::New(buf);
  memcpy(clone, foo, sizeof(*foo));

  clone->set_bar(mojo::internal::Clone(foo->bar(), buf));
  clone->set_data(mojo::internal::Clone(foo->data(), buf));
  clone->set_extra_bars(mojo::internal::Clone(foo->extra_bars(), buf));
  clone->set_name(mojo::internal::Clone(foo->name(), buf));
  clone->set_files(mojo::internal::Clone(foo->files(), buf));

  return clone;
}

// static
void ObjectTraits<sample::Foo>::EncodePointersAndHandles(
    sample::Foo* foo, std::vector<Handle>* handles) {
  Encode(&foo->bar_, handles);
  Encode(&foo->data_, handles);
  Encode(&foo->extra_bars_, handles);
  Encode(&foo->name_, handles);
  Encode(&foo->files_, handles);
}

// static
bool ObjectTraits<sample::Foo>::DecodePointersAndHandles(
    sample::Foo* foo, const Message& message) {
  if (!Decode(&foo->bar_, message))
    return false;
  if (!Decode(&foo->data_, message))
    return false;
  if (foo->_header_.num_fields >= 8) {
    if (!Decode(&foo->extra_bars_, message))
      return false;
  }
  if (foo->_header_.num_fields >= 9) {
    if (!Decode(&foo->name_, message))
      return false;
  }
  if (foo->_header_.num_fields >= 10) {
    if (!Decode(&foo->files_, message))
      return false;
  }

  // TODO: validate
  return true;
}

template <>
class ObjectTraits<sample::Service_Frobinate_Params> {
 public:
  static void EncodePointersAndHandles(
      sample::Service_Frobinate_Params* params,
      std::vector<Handle>* handles) {
    Encode(&params->foo_, handles);
    EncodeHandle(&params->port_, handles);
  }

  static bool DecodePointersAndHandles(
      sample::Service_Frobinate_Params* params,
      const Message& message){
    if (!Decode(&params->foo_, message))
      return false;
    if (params->_header_.num_fields >= 3) {
      if (!DecodeHandle(&params->port_, message.handles))
        return false;
    }

    // TODO: validate
    return true;
  }
};

template <>
class ObjectTraits<sample::ServiceClient_DidFrobinate_Params> {
 public:
  static void EncodePointersAndHandles(
      sample::ServiceClient_DidFrobinate_Params* params,
      std::vector<Handle>* handles) {
  }

  static bool DecodePointersAndHandles(
      sample::ServiceClient_DidFrobinate_Params* params,
      const Message& message) {
    return true;
  }
};

}  // namespace internal
}  // namespace mojo

File: mojo/public/bindings/sample/mojom/sample_service.js
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

define([
    "mojo/public/bindings/js/core",
    "mojo/public/bindings/js/codec",
  ], function(core, codec) {

  function Bar() {
    this.alpha = 0;
    this.beta = 0;
    this.gamma = 0;
  }

  Bar.encodedSize = codec.kStructHeaderSize + 8;

  Bar.decode = function(decoder) {
    var packed;
    var val = new Bar();
    var numberOfBytes = decoder.read32();
    var numberOfFields = decoder.read32();
    val.alpha = decoder.read8();
    val.beta = decoder.read8();
    val.gamma = decoder.read8();
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    return val;
  };

  Bar.encode = function(encoder, val) {
    var packed;
    encoder.write32(Bar.encodedSize);
    encoder.write32(3);
    encoder.write8(val.alpha);
    encoder.write8(val.beta);
    encoder.write8(val.gamma);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
  };

  function Foo() {
    this.x = 0;
    this.y = 0;
    this.a = false;
    this.b = false;
    this.c = false;
    this.bar = null;
    this.data = [];
    this.extra_bars = [];
    this.name = "";
    this.files = [];
  }

  Foo.encodedSize = codec.kStructHeaderSize + 56;

  Foo.decode = function(decoder) {
    var packed;
    var val = new Foo();
    var numberOfBytes = decoder.read32();
    var numberOfFields = decoder.read32();
    val.x = decoder.read32();
    val.y = decoder.read32();
    packed = decoder.read8();
    val.a = (packed >> 0) & 1 ? true : false;
    val.b = (packed >> 1) & 1 ? true : false;
    val.c = (packed >> 2) & 1 ? true : false;
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    val.bar = decoder.decodeStructPointer(Bar);
    val.data = decoder.decodeArrayPointer(codec.Uint8);
    val.extra_bars = decoder.decodeArrayPointer(new codec.PointerTo(Bar));
    val.name = decoder.decodeStringPointer();
    val.files = decoder.decodeArrayPointer(codec.Handle);
    return val;
  };

  Foo.encode = function(encoder, val) {
    var packed;
    encoder.write32(Foo.encodedSize);
    encoder.write32(10);
    encoder.write32(val.x);
    encoder.write32(val.y);
    packed = 0;
    packed |= (val.a & 1) << 0
    packed |= (val.b & 1) << 1
    packed |= (val.c & 1) << 2
    encoder.write8(packed);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.encodeStructPointer(Bar, val.bar);
    encoder.encodeArrayPointer(codec.Uint8, val.data);
    encoder.encodeArrayPointer(new codec.PointerTo(Bar), val.extra_bars);
    encoder.encodeStringPointer(val.name);
    encoder.encodeArrayPointer(codec.Handle, val.files);
  };

  function Service_Frobinate_Params() {
    this.foo = null;
    this.baz = false;
    this.port = codec.kInvalidHandle;
  }

  Service_Frobinate_Params.encodedSize = codec.kStructHeaderSize + 16;

  Service_Frobinate_Params.decode = function(decoder) {
    var packed;
    var val = new Service_Frobinate_Params();
    var numberOfBytes = decoder.read32();
    var numberOfFields = decoder.read32();
    val.foo = decoder.decodeStructPointer(Foo);
    val.baz = decoder.read8() & 1;
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    val.port = decoder.decodeHandle();
    return val;
  };

  Service_Frobinate_Params.encode = function(encoder, val) {
    var packed;
    encoder.write32(Service_Frobinate_Params.encodedSize);
    encoder.write32(3);
    encoder.encodeStructPointer(Foo, val.foo);
    encoder.write8(1 & val.baz);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.encodeHandle(val.port);
  };

  function ServiceClient_DidFrobinate_Params() {
    this.result = 0;
  }

  ServiceClient_DidFrobinate_Params.encodedSize = codec.kStructHeaderSize + 8;

  ServiceClient_DidFrobinate_Params.decode = function(decoder) {
    var packed;
    var val = new ServiceClient_DidFrobinate_Params();
    var numberOfBytes = decoder.read32();
    var numberOfFields = decoder.read32();
    val.result = decoder.read32();
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    decoder.skip(1);
    return val;
  };

  ServiceClient_DidFrobinate_Params.encode = function(encoder, val) {
    var packed;
    encoder.write32(ServiceClient_DidFrobinate_Params.encodedSize);
    encoder.write32(1);
    encoder.write32(val.result);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
    encoder.skip(1);
  };
  var kService_Frobinate_Name = 0;

  function ServiceProxy(receiver) {
    this.receiver_ = receiver;
  }
  ServiceProxy.prototype.frobinate = function(foo, baz, port) {
    var params = new Service_Frobinate_Params();
    params.foo = foo;
    params.baz = baz;
    params.port = port;

    var builder = new codec.MessageBuilder(
        kService_Frobinate_Name,
        codec.align(Service_Frobinate_Params.encodedSize));
    builder.encodeStruct(Service_Frobinate_Params, params);
    var message = builder.finish();
    this.receiver_.accept(message);
  };

  function ServiceStub() {
  }

  ServiceStub.prototype.accept = function(message) {
    var reader = new codec.MessageReader(message);
    switch (reader.messageName) {
    case kService_Frobinate_Name:
      var params = reader.decodeStruct(Service_Frobinate_Params);
      this.frobinate(params.foo, params.baz, params.port);
      return true;
    default:
      return false;
    }
  };
  var kServiceClient_DidFrobinate_Name = 0;

  function ServiceClientProxy(receiver) {
    this.receiver_ = receiver;
  }
  ServiceClientProxy.prototype.didFrobinate = function(result) {
    var params = new ServiceClient_DidFrobinate_Params();
    params.result = result;

    var builder = new codec.MessageBuilder(
        kServiceClient_DidFrobinate_Name,
        codec.align(ServiceClient_DidFrobinate_Params.encodedSize));
    builder.encodeStruct(ServiceClient_DidFrobinate_Params, params);
    var message = builder.finish();
    this.receiver_.accept(message);
  };

  function ServiceClientStub() {
  }

  ServiceClientStub.prototype.accept = function(message) {
    var reader = new codec.MessageReader(message);
    switch (reader.messageName) {
    case kServiceClient_DidFrobinate_Name:
      var params = reader.decodeStruct(ServiceClient_DidFrobinate_Params);
      this.didFrobinate(params.result);
      return true;
    default:
      return false;
    }
  };

  var exports = {};
  exports.Bar = Bar;
  exports.Foo = Foo;
  exports.ServiceProxy = ServiceProxy;
  exports.ServiceStub = ServiceStub;
  exports.ServiceClientProxy = ServiceClientProxy;
  exports.ServiceClientStub = ServiceClientStub;
  return exports;
});
